### PR TITLE
Binary search to minimize number of context lines

### PR DESCRIPTION
**Pull request by akaihola**
If the AST differs after reformatting when diffing and chunking with zero context lines, search for the lowest successful number of context lines using a binary search instead of by incrementing one by one. This vastly improves performance in such problematic cases.

### CODE REVIEW

File: src/darker/__main__.py [hunk 1]
```diff
@@ -121,22 +130,27 @@ def format_edited_parts(
...
-                    "AST verification failed. "
-                    "Trying again with %s lines of context for `git diff -U`",
-                    context_lines + 1,
+                logger.warning(
+                    "AST verification of %s with %s lines of context failed",
+                    src,
+                    context_lines,
                 )
-                continue
+                minimum_context_lines.respond(False)
```
Comment (samoylovfp) [inline src/darker/__main__.py:138]: :star_struck: love the BinarySearch API!

File: src/darker/verification.py [hunk 2]
```diff
@@ -11,6 +11,50 @@ class NotEquivalentError(Exception):
...
+    """Effectively search the first index for which a condition is ``True``
+
+    Example usage to find first number between 0 and 100 whose square is larger than
+    1000::
+
+        s = BinarySearch(0, 100)
+        while not s.found:
+            s.respond(s.get_next() ** 2 > 1000)
+        answer = s.result
+
```
Comment (samoylovfp) [inline src/darker/verification.py:24]: Maybe make it a doctest with an assertion?
```suggestion
        assert s.result == 32
```

File: src/darker/__main__.py [hunk 3]
```diff
@@ -79,7 +79,16 @@ def format_edited_parts(
         else:
             edited = worktree_content
         max_context_lines = len(edited.lines)
-        for context_lines in range(max_context_lines + 1):
+        minimum_context_lines = BinarySearch(0, max_context_lines + 1)
+        last_successful_reformat = None
+        while not minimum_context_lines.found:
+            context_lines = minimum_context_lines.get_next()
+            if context_lines > 0:
+                logger.warning(
```
Comment (samoylovfp) [inline src/darker/__main__.py:87]: Was this changed to a warning for debugging purposes?
Looks like `debug` was the appropriate level

File: src/darker/tests/test_verification.py [hunk 4]
```diff
@@ -30,3 +29,23 @@ def test_verify_ast_unchanged(src_content, dst_content, expect):
...
+    """``darker.verification.BinarySearch``"""
+    search = BinarySearch(0, 5)
+    tries = []
+    while not search.found:
+        tries.append(search.get_next())
+
+        search.respond(tries[-1] > 2)
+
+    assert search.result == 3
+    assert tries == [0, 3, 2]
```
Comment (samoylovfp) [inline src/darker/tests/test_verification.py:51]: Sort-of fuzzy-testing
```suggestion
    assert tries == [0, 3, 2]


@pytest.mark.parametrize("i", range(50))
def test_binary_search_in_50(i):
    b = BinarySearch(0, 50)
    while not b.found:
        b.respond(b.get_next() >= i)
    assert b.result == i

```

**Review by samoylovfp — APPROVED** [review]
Looks awesome! A couple of nit-picks in comments

**Comment by akaihola** [comment]
Thanks @samoylovfp for the review! This made my already decent morning perfect!

**Pull request merged**